### ISSUE TITLE
Bash programmable completion

### ISSUE THREAD

**Issue by draffelt** [MRtrix3/mrtrix3#267]
Ok, so this one is definitely on the wish list, however I thought I would put it up in case some enthusiastic mrtrix user is bored on a rainy day.

It would be nice to have a script that parses the current commands and generates a bash completion definitions file. So that when we hit tab typing the name of a command, it will only list the expected input file types. It would auto complete command options, which would be handy if you know roughly what you want, but not the exact option (for example -seed_* in tckgen). 

See here: http://www.gnu.org/software/bash/manual/html_node/Programmable-Completion.html

For an example see git: /etc/bash_completion.d/git-prompt

**Comment by draffelt** [comment]
Another example here: http://fahdshariff.blogspot.com.au/2011/04/writing-your-own-bash-completion.html

**Comment by jdtournier** [comment]
:+1: would indeed be nice...

**Comment by rtabbara** [comment]
First attempt of this is now in <code>bash_completion</code>, which allows for bash completion support for mrtrix command options.  Unfortunately, trying to find a map between commands/options to expected file types proved too difficult --- I think additional meta-data within the command's help pages is probably required in order to accurately determine this.

There's now two files in the scripts directory:

* generate_mrtrix_cmd_completion.sh
* mrtrix_command_completion

The second file is the pre-generated output of the first script &mdash; it's this static file that bash needs to know about to get auto-completion working. Given that generating the completion file is reliant on there being a pre-existing installation of mrtrix, I think using the static completion file might be easier for deployment (i.e. something that can be referenced in <code>configure</code>).

Anyway, happy to sit on this until it's more fleshed out. Just in case anyone's interested...

**Comment by jdtournier** [comment]
In a previous life, I thought it would be nice to add a GUI to MRtrix commands, in the form of a wrapper script that would figure out what forms to present to the user dynamically for each command. The idea was to first run each command with a single special argument that would cause it to dump its usage information to stdout in an easy to parse format, and the script would build the GUI from that information. I'd gotten as far as adding the option to dump the usage info, but obviously never got further than that. But it's still in the code...

Have a go at running each command with the `__print_full_usage__` argument. This invokes the `App::full_usage ()` function, defined in `lib/app.cpp`, line 180. Hopefully that'll fit the bill. If it doesn't, feel free to modify it as needed.

By the way, that's the same approach as used to update the wiki information for each command, although via a different function specifically to output markdown.

**Comment by draffelt** [comment]
Nice work Rami. Apparently dreams do come true. I really like the fact that I can type "-" after a command then tab to quickly see all options. It would be the icing on the cake if you can get the tab to list only expected file types within the directory. I'm guessing it might be easy for each option, however for the required arguments I'm not sure if you can specify the file types for each specific argument number (though I have not looked into this).

**Comment by rtabbara** [comment]
A new update of this script has been added to the <code>bash_completion</code> branch. This now parses each command's `__print_full_usage__` output and generates a completion file <code>mrtrix_command_completion</code> that now also supports completion of command/option arguments.

There's a few finicky parsing issues to contend with such as non-adjacent command arguments, so any robustness tests would be helpful.

**Comment by rtabbara** [comment]
Was also discussing with Dave, but interested in suggestions on how best to deploy the completion file once it seems to be working ok. e.g. Should it be part of the build script? Should we bundle a precomputed version that's added when configuring.

FWIW, it takes about 2 seconds on my machine to currently run the script.

**Comment by draffelt** [comment]
To add to that. I was thinking that we could run it upon successful
completion of the build script. This will ensure that when a user updates,
any command changes are incorporated into the bash completion.

We could possibly have an option to disable it in the configure script, for
developers who build MRtrix 100 times a day and don't want to wait an extra
2 seconds each time.

By the way, the script outputs the bash completion file to the scripts
folder, which can be then loaded by adding a line to ~/.bashrc

**Comment by Lestropie** [comment]
Given not everyone uses Bash, and some have multiple config files for building release / debug in parallel, I don't think automating its generation as part of the build is the right move. It would be a lot of redundant generation.

I'd commit the generating script, which can be run manually at any time; but also have the resulting completion file included in the repo. That way, whenever a commit introduces a new command or modifies command-line arguments / options, the committer _should_ run the script so that the changes are reflected; or it can be run and the updated file committed after the fact; or users can run it themselves.

**Comment by draffelt** [comment]
Yeah we were thinking that we could force the committer to do it. Initially
I was worried people might forget to run it. However, now I think about it
it's probably not a bad way to go. In case people forget, we could have a
script on the mrtrix.org website server that updates and runs it daily then
commits if the file changes.

**Comment by jdtournier** [comment]
Good work Rami! I reckon that'll make a lot more people happy than just @draffelt - although he's clearly pretty stoked about it...

My personal preference is to keep _generated_ files out of the repo - it should really only contain the source files. Having to update the bash completion file _after_ the fact on the many occasions when the developer will forget is just going to clutter up the history... 

I think it should be fairly trivial to extend the script to check the timestamps between the executables and the generated bash completion file, so that it would only run when required (you can always `touch bin/mrinfo` or something to force a re-run of the script). You could even set up a Makefile for it... That way those developers who really want it can run `./build && ./scripts/generate_mrtrix_cmd_completion.sh`. Or we can add an option like `-with-bash-completion` to the build script to additionally run that script if needed (probably the simplest - bash is pretty much the default on all platforms, even Windows - but I agree that's no reason to force the script to run every time). Or developers can set up an alias like `alias ./build='./build && ./scripts/generate_mrtrix_cmd_completion.sh'` to get it to run by default on their system. Or any number of other alternative solutions to achieve the same thing.

In the long run, we'll need to add an `-install` target to the build script, and at that point we'll get the script to run - along with installing desktop files and icons, etc. 

Anyway, that's how I'd prefer to see things set up. Does that sound reasonable?

In the meantime, why not merge this branch, since it has no impact on the rest of the code - it's not like editing this script on master is going to cause any other command to fail...

**Comment by jdtournier** [comment]
Sorry, hit the wrong button...

**Comment by draffelt** [comment]
Given that most systems will use bash (which we can check for anyway), I'm thinking we should run this script automatically at the end of the build script if everything succeeds. As you suggested we can check the timestamp of all cmd/*cpp files and only run it if one file is newer than the generated file.

Having it run automatically is good since most users (including me), will forget to run it when something is updated. 

We can also add a line to the README.md file (right after where we tell them to set up the PATH), which tells users to add the appropriate line to their .bashrc file.  If we don't help lay users set this up (and hide it as an option) then most users will miss it. 

Agreed that we later once we have an -install option the generated file should go to /etc/bash_completion.d/

**Comment by jdtournier** [comment]
I'd be OK with that - it is true that bash is the default on all target systems. I think eventually this step will move to the `install` target anyway, so we should consider this temporary until we get this set up. 

For the record, the timestamps to check are for all `bin/*` files... 

Is the current script to generate the completion files written in python? If so it should be trivial to write a function for it, and set up a few rules to match the bash completion files. We can then set them as the default targets, which will bring in all the executables along with them... Happy to help if you want.

**Comment by rtabbara** [comment]
OK, I've ported the generator script from bash to python mainly to make things a bit neater.

@jdtournier: I think you were suggesting to associate each completion file to each corresponding target command, but currently a single completion file is generated that handles all commands. Given that it takes such a short amount of time (~2 sec.) to run, I'm not sure whether there would be a benefit to partitioning this up.

I've also modified the build script to run the bash completion script at the end of a successful build if it's possible and necessary. In order to use the generated completion file, a user will need to add something like the following to .bashrc

<pre>
<code>
if [ -f my_mrtrix_dir/scripts/mrtrix_bash_completion ]; then
	. my_mrtrix_dir/scripts/mrtrix_bash_completion
fi
</code>
</pre>

Given that this change touches the build script, I was reluctant to push this directly into master and currently I'm still on the bash_completion branch. If people are ok with this being merged, then please let me know.

**Comment by jdtournier** [comment]
Why don't we merge this into the `updated_syntax` branch? That way the changes would remain out of master while we check it out, and most of us are or will soon be using that branch full time anyway...

**Comment by rtabbara** [comment]
@jdtournier Sounds good, and it has been done.

**Comment by jdtournier** [comment]
:+1:

Good to close then?

**Comment by draffelt** [comment]
Working well so far for me! :D

**Comment by jdtournier** [comment]
Just thought I'd mention that the `generate_mrtrix_cmd_completion.py` script only works with Python 2 - doesn't produce usable output with Python 3 - took me a little while to figure out why it wasn't working for me...

**Comment by rtabbara** [comment]
@jdtournier Thanks for picking this us up. Should be fixed now. FWIW,  in Python 3+, I had to explicitly decode my subprocess' stdout so that it was treated as a string.

**Comment by jdtournier** [comment]
Yes, I remember the fun I had updating the build and configure scripts to work under Python 3... All works now, thanks for the quick fix!
